### ISSUE TITLE
Plugin portal should support version listing

### ISSUE THREAD

**Issue by IonutNegru87** [gradle/gradle#1297]
The latest version for an gradle plugin is not detected correctly. The issue I'm having is specific to **Gradle ErrorProne Plugin**, but this could also happen with other 3rd party libraries/plugins.

### Expected Behavior
The latest version detected should be **0.0.8**

### Current Behavior
The latest version is not detected, instead an old version is detected(**0.0.6**).
This is the last version deployed on Central.

This could be due to the fact that `maven-metadata.xml` isn't queried correctly or it isn't updated when a new version is deployed.

### Context
I'm using [gradle-versions-plugin](https://github.com/ben-manes/gradle-versions-plugin) to detect when new versions of the dependencies of my project are available. Everything works correctly except for the [gradle-errorprone-plugin](https://github.com/tbroyer/gradle-errorprone-plugin).
Conform latest documentation and gradle plugin info, the latest version is **0.0.8**, but the plugin detects that the latest version is **0.0.6**:
> - net.ltgt.gradle:gradle-errorprone-plugin [0.0.8 <- 0.0.6]

After opening an bug for that plugin "[Plugin latest version not detected](https://github.com/ben-manes/gradle-versions-plugin/issues/126)", it was explained to me that the issue resides in how gradle reads the metadata for an plugin/library.

### Steps to Reproduce (for bugs)
Integrate **gradle-versions-plugin**
Integrate **gradle-errorprone-plugin**
Disallow release candidates as upgradable versions by adding the following strategy at the build.gradle of the root of the project:
```
dependencyUpdates.resolutionStrategy = {
    componentSelection { rules ->
        rules.all { ComponentSelection selection ->
            boolean rejected = ['alpha', 'beta', 'rc', 'cr', 'm'].any { qualifier ->
                selection.candidate.version ==~ /(?i).*[.-]${qualifier}[.\d-]*/
            }
            if (rejected) {
                selection.reject('Release candidate')
            }
        }
    }
}
```
Run the gradle task for dependency updates:
`gradlew dependencyUpdates'

### Your Environment

- Android Studio 2.2.3 (stable release)
- Gradle 3.3 (wrapper inside project)
- Android Gradle plugin 2.2.3
- Java 1.8.0_111

**Comment by ben-manes** [comment]
Some of the prior requests:
 - [plugin portal's maven-metadata.xml](https://groups.google.com/forum/#!searchin/gradle-dev/ben$20manes%7Csort:relevance/gradle-dev/0E1khrWRIbo/Zc76Usu8AgAJ)
- [Finding a plugins latest version programmatically?](https://discuss.gradle.org/t/finding-a-plugins-latest-version-programmatically/9986)
 - [Why the Gradle Central Plugin Repository does not support dynamic plugin versions?](https://discuss.gradle.org/t/why-the-gradle-central-plugin-repository-does-not-support-dynamic-plugin-versions/11788) @szczepiq

**Comment by ben-manes** [comment]
@aalmiray [reported](https://twitter.com/aalmiray/status/875687365393735681) this on Twitter regarding Griffin (@alkemist, @eriwen, @melix). The plugin repository not handling metadata properly means the latest version of the `build-scan` plugin (and others) cannot be determined.

**Comment by aalmiray** [comment]
To be fair, it's not just the plugin portal at fault. The Griffon build has trouble with other dependencies that may not have published accurate date, such as the following report shows

	Failed to determine the latest version for the following dependencies (use --info for details):
	 - com.gradle:build-scan-plugin
	 - org.easytesting:fest-swing-junit-4.5
	 - javax.inject:javax.inject
	 - gradle.plugin.net.ossindex:ossindex-gradle-plugin
	 - com.gradle.publish:plugin-publish-plugin
	 - org.testfx:testfx-core
	 - org.testfx:testfx-junit

**Comment by ben-manes** [comment]
You would have to look at the details. A quick glance,
 - The plugins are due to this issue.
 - TestFX only has `alpha` releases on Central, which your [resolution rule](https://github.com/griffon/griffon/blob/development/build.gradle#L94) rejects.
 - The rest aren't clear, but the `info` log should help clarify

The versions plugin is a script on top of Gradle's resolution logic using dynamic version support and its internal comparator. The report should be the same as if you used `+` for all your dependencies. This makes it simple, consistent, and fewer surprises as it integrates with your other build logic (e.g. an S3 repository). The only flaw is being blocked by this issue, as I'd prefer not hacking around it by scraping the plugin portal's html....

**Comment by aalmiray** [comment]
Indeed. I'm aware that other dependencies may not have published correct data (which is the case of the FEST related artifacts). I was just providing additional data that might help as use cases. Frankly I've no idea why some external plugins developed by Gradle (such as build-scan) appear to have incorrect or missing metadata at their own portal.

**Comment by ben-manes** [comment]
I checked out Griffon and looked at the info log. The `fest-swing-junit-4.5` and `javax.inject:javax.inject` dependencies are only resolved using jcenter and, for some reason, the metadata isn't published there (nor does it proxy to Central to find it). If you add Central then that should resolve those two.

It would be very nice if alternatives to Maven Center used compatible repository configurations. Unfortunately both JCenter and the plugin portal have quirks that can cause resolution issues.

**Comment by oehme** [comment]
Duplicate of #1050

**Comment by ben-manes** [comment]
That does not appear to be a duplicate. 

This task highlights that the Gradle Plugin Portal is __not__ a compliant m2 repository. It has stale metadata and rejects required HTTP requests used for querying. This has nothing to do with whether some dependencies use the plugin marker artifacts format as coordinates in the buildscript syntax fail as well. As stated in that issue, users are currently proxying using buildscript syntax due to internal restrictions.

#1050 states its fix is data rewriting of the coordinates and updating to Gradle's publishing plugin. The portal allows point queries of a target version but does not allow range queries. The maven metadata is also incorrect, as it is (was?) written only when a plugin is created but not on subsequent releases.

This issue requests that the portal be a fully compatible m2 repository and that publishing updates the maven metadata. This would allow Gradle to perform a dynamic version query to determine the latest version. For the versions plugin this would allow a report to show the upgrade status and for others, like Netflix's nebula, to update a lock files instead.

**Comment by oehme** [comment]
The other issue includes making version listing work, otherwise the feature expressed in the title wouldn't be complete. I've adjusted the description to reflect that.

**Comment by ben-manes** [comment]
Thanks!

**Comment by ben-manes** [comment]
This seems to be mostly fixed by Gradle special casing plugins during metadata resolution.

Any newish plugin has `gradle.plugin` in the group name. This is then resolved to the latest version.

Older plugins that do not, such as the ErrorProne and the JMH plugins, are resolved using `maven-metadata.xml`. However, the plugin portal does not contain those files and the lookup fails. Since the plugins were previously on other repositories (Central, JCenter) the metadata is returns a stale version.

Thanks for the partial fix. Can we have support for older plugins?

**Comment by oehme** [comment]
@ben-manes If I understand your request correctly, this should be fixed by https://github.com/gradle/gradle/pull/3347. Gradle will no longer use the magic portal protocol, but instead look for plugin marker POMs. These are synthesized by the portal for both new and old versions. As a result, your versions plugin should correctly inform the user about newer versions.

**Comment by eriwen** [comment]
I forgot to update this issue as well that one can get all plugin versions, including the latest, through maven-metadata.xml files on the Gradle plugin portal now, as #3347 has been completed (and #1050 has been delivered)

Does this fulfill everyone's needs?

**Comment by ben-manes** [comment]
Yes, thank you!

It looks like you recently fixed the quirk with [plugin marker artifacts](https://docs.gradle.org/current/userguide/plugins.html#sec:plugin_markers), which was causing declarations like `net.ltgt.gradle:gradle-errorprone-plugin:$version` to not be resolved. Previously one had to expand it to the form `net.ltgt.errorprone:net.ltgt.errorprone.gradle.plugin:$version` for the metadata query to work.
